# [Project - Anomaly Detection](https://classroom.google.com/u/1/c/NTI2MzQ4Nzc3NTk3/a/NTI2MzQ4Nzc3NjUx/details)

In [1]:
#################################################
#################### Imports ####################
#################################################

# ---------------- #
# Common Libraries #
# ---------------- #
      
# Standard Imports
import os
import requests
import numpy as np
import pandas as pd

# Working with Dates & Times
from sklearn.model_selection import TimeSeriesSplit
from datetime import timedelta, datetime

# Working with Math & Stats
import statsmodels.api as sm
import scipy.stats as stats

# to evaluate performance using rmse
from sklearn.metrics import mean_squared_error
from math import sqrt 

# holt's linear trend model. 
from statsmodels.tsa.api import Holt

# Plots, Graphs, & Visualization
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib.ticker import StrMethodFormatter
from matplotlib.dates import DateFormatter

# plotting defaults
plt.rc('figure', figsize=(13, 7))
plt.style.use('seaborn-whitegrid')
plt.rc('font', size=16)

# -------------- #
# Action Imports #
# -------------- #

# Warnings 
import warnings 
warnings.filterwarnings("ignore")

# ------------ #
# JUPYTER ONLY #
# ------------ #

# Left Align Tables in Jupyter Notebook
from IPython.core.display import HTML

# Disable autosave
%autosave 0

# Increases Display Resolution for Graphs 
%matplotlib inline 
%config InlineBackend.figure_format = 'retina'

# ------------- #
# Local Imports #
# ------------- #

# importing sys
import sys

# adding 00_helper_files to the system path
sys.path.insert(0, '/Users/qmcbt/codeup-data-science/00_helper_files')

# env containing sensitive access credentials
import env
from env import user, password, host

# Import Helper Modules
import QMCBT_00_quicktips as qt
import QMCBT_01_acquire as acq
import QMCBT_02_prepare as prep
import QMCBT_03_explore as exp
import QMCBT_04_visualize as viz
import QMCBT_05_model as mod
import QMCBT_wrangle as qw
import wrangle as w

Autosave disabled


In [2]:
# Left Align Tables in Jupyter Notebook
table_css = 'table {align:left;display:block}'
HTML('{}'.format(table_css))

# SYLABUS
* To get 100 on this project you only need to answer 5 out of the 7 questions (along with the other deliverables listed below i.e. slide, your notebook, etc).
* send your email before the due date and time to datascience@codeup.com (Only one team member can do this on behalf of whole team).
* Submit a link to a final notebook on GitHub that asks and answers questions - document the work you do to justify findings
* Compose an email with the answers to the questions/your findings, and in the email, include the link to your notebook in GitHub and attach your slide.
* You will not present this, so be sure that the details you need your leader to convey/understand are clearly communicated in the email.
* Your slide should be like an executive summary and be in form to present.
* Continue to use best practices of acquire.py, prepare.py, etc.
* Since there is no modeling to be done for this project, there is no need to split the data into train/validate/test
* The cohort schedule is in the SQL database, and alumni.codeup.com has info as well.
* Teamwork with Git handout is posted in the google classroom

I have some questions for you that I need to be answered before the board meeting Wednesday afternoon. I need to be able to speak to the following questions. I also need a single slide that I can incorporate into my existing presentation (Google Slides) that summarizes the most important points. My questions are listed below; however, if you discover anything else important that I didn’t think to ask, please include that as well.

# Download the data
url = 'https://drive.google.com/u/1/uc?id=1phD962Wrt8fetpvX-ersybPcZW3_54ma&export=download'

In [3]:
df = w.acquire_anonymized_curriculum_access_data()

In [4]:
df = w.clean_the_data(df)

In [5]:
df.T

date                 2018-01-26           2018-01-26  \
date        2018-01-26 00:00:00  2018-01-26 00:00:00   
time                   09:55:03             09:56:02   
page                          /              java-ii   
user_id                       1                    1   
cohort_id                   8.0                  8.0   
ip                 97.105.19.61         97.105.19.61   
name                    Hampton              Hampton   
start_date  2015-09-22 00:00:00  2015-09-22 00:00:00   
end_date    2016-02-06 00:00:00  2016-02-06 00:00:00   
program_id                  PHP                  PHP   

date                                 2018-01-26  \
date                        2018-01-26 00:00:00   
time                                   09:56:05   
page        java-ii/object-oriented-programming   
user_id                                       1   
cohort_id                                   8.0   
ip                                 97.105.19.61   
name                                    Hampton   
start_date                  2015-09-22 00:00:00   
end_date                    2016-02-06 00:00:00   
program_id                                  PHP   

date                                2018-01-26                 2018-01-26  \
date                       2018-01-26 00:00:00        2018-01-26 00:00:00   
time                                  09:56:06                   09:56:24   
page        slides/object_oriented_programming  javascript-i/conditionals   
user_id                                      1                          2   
cohort_id                                  8.0                       22.0   
ip                                97.105.19.61               97.105.19.61   
name                                   Hampton                      Teddy   
start_date                 2015-09-22 00:00:00        2018-01-08 00:00:00   
end_date                   2016-02-06 00:00:00        2018-05-17 00:00:00   
program_id                                 PHP                       Java   

date                 2018-01-26                 2018-01-26  \
date        2018-01-26 00:00:00        2018-01-26 00:00:00   
time                   09:56:41                   09:56:46   
page         javascript-i/loops  javascript-i/conditionals   
user_id                       2                          3   
cohort_id                  22.0                       22.0   
ip                 97.105.19.61               97.105.19.61   
name                      Teddy                      Teddy   
start_date  2018-01-08 00:00:00        2018-01-08 00:00:00   
end_date    2018-05-17 00:00:00        2018-05-17 00:00:00   
program_id                 Java                       Java   

date                    2018-01-26           2018-01-26  \
date           2018-01-26 00:00:00  2018-01-26 00:00:00   
time                      09:56:48             09:56:59   
page        javascript-i/functions   javascript-i/loops   
user_id                          3                    2   
cohort_id                     22.0                 22.0   
ip                    97.105.19.61         97.105.19.61   
name                         Teddy                Teddy   
start_date     2018-01-08 00:00:00  2018-01-08 00:00:00   
end_date       2018-05-17 00:00:00  2018-05-17 00:00:00   
program_id                    Java                 Java   

date                    2018-01-26  ...           2021-04-21  \
date           2018-01-26 00:00:00  ...  2021-04-21 00:00:00   
time                      09:58:26  ...             16:38:14   
page        javascript-i/functions  ...    java-iii/servlets   
user_id                          4  ...                  834   
cohort_id                     22.0  ...                134.0   
ip                    97.105.19.61  ...          67.11.50.23   
name                         Teddy  ...                 Luna   
start_date     2018-01-08 00:00:00  ...  2020-12-07 00:00:00   
end_date       2018-05-17 00:00:00  ...  2021-06-08 00:00:00   


## 1. Which lesson appears to attract the most traffic consistently across cohorts (per program)?

## 2. Is there a cohort that referred to a lesson significantly more than other cohorts seemed to gloss over?

## 3. Are there students who, when active, hardly access the curriculum? If so, what information do you have about these students?

## 4. Is there any suspicious activity, such as users/machines/etc accessing the curriculum who shouldn’t be? Does it appear that any web-scraping is happening? Are there any suspicious IP addresses?

## 5. At some point in 2019, the ability for students and alumni to access both curriculums (web dev to ds, ds to web dev) should have been shut off. Do you see any evidence of that happening? Did it happen before?

<div class="alert alert-warning">

# 6. What topics are grads continuing to reference after graduation and into their jobs (for each program)?

# Grad Users

In [6]:
# Create DataFrame for users that have accessed the curriculum past there graduation date
grad_users = df[df.date > df.end_date]

In [7]:
grad_users.T

date                 2018-01-26           2018-01-26  \
date        2018-01-26 00:00:00  2018-01-26 00:00:00   
time                   09:55:03             09:56:02   
page                          /              java-ii   
user_id                       1                    1   
cohort_id                   8.0                  8.0   
ip                 97.105.19.61         97.105.19.61   
name                    Hampton              Hampton   
start_date  2015-09-22 00:00:00  2015-09-22 00:00:00   
end_date    2016-02-06 00:00:00  2016-02-06 00:00:00   
program_id                  PHP                  PHP   

date                                 2018-01-26  \
date                        2018-01-26 00:00:00   
time                                   09:56:05   
page        java-ii/object-oriented-programming   
user_id                                       1   
cohort_id                                   8.0   
ip                                 97.105.19.61   
name                                    Hampton   
start_date                  2015-09-22 00:00:00   
end_date                    2016-02-06 00:00:00   
program_id                                  PHP   

date                                2018-01-26           2018-01-26  \
date                       2018-01-26 00:00:00  2018-01-26 00:00:00   
time                                  09:56:06             10:14:47   
page        slides/object_oriented_programming                    /   
user_id                                      1                   11   
cohort_id                                  8.0                  1.0   
ip                                97.105.19.61         97.105.19.61   
name                                   Hampton               Arches   
start_date                 2015-09-22 00:00:00  2014-02-04 00:00:00   
end_date                   2016-02-06 00:00:00  2014-04-22 00:00:00   
program_id                                 PHP                  PHP   

date                 2018-01-26              2018-01-26  \
date        2018-01-26 00:00:00     2018-01-26 00:00:00   
time                   10:14:51                10:15:02   
page               javascript-i  javascript-i/functions   
user_id                      11                      11   
cohort_id                   1.0                     1.0   
ip                 97.105.19.61            97.105.19.61   
name                     Arches                  Arches   
start_date  2014-02-04 00:00:00     2014-02-04 00:00:00   
end_date    2014-04-22 00:00:00     2014-04-22 00:00:00   
program_id                  PHP                     PHP   

date                    2018-01-26           2018-01-26           2018-01-26  \
date           2018-01-26 00:00:00  2018-01-26 00:00:00  2018-01-26 00:00:00   
time                      10:40:15             10:59:02             10:59:06   
page        javascript-i/functions           index.html               java-i   
user_id                          1                   21                   21   
cohort_id                      8.0                 16.0                 16.0   
ip                    97.105.19.61        71.137.146.71        71.137.146.71   
name                       Hampton              Niagara              Niagara   
start_date     2015-09-22 00:00:00  2016-10-26 00:00:00  2016-10-26 00:00:00   
end_date       2016-02-06 00:00:00  2017-03-09 00:00:00  2017-03-09 00:00:00   
program_id                     PHP                 Java                 Java   

date        ...                 2021-04-21           2021-04-21  \
date        ...        2021-04-21 00:00:00  2021-04-21 00:00:00   
time        ...                   16:07:17             16:30:30   
page        ...  javascript-ii/RESTful-api             appendix   
user_id     ...                         64                  771   
cohort_id   ...                       28.0                 62.0   
ip          ...              71.150.217.33      172.125.226.175   
name        ...                      St

In [8]:
# Count the users
grad_users.user_id.nunique()

592

In [9]:
# Show the users
grad_users.user_id.unique()

array([  1,  11,  21,  26,  37,  40,  50,  51,  53,  60,  64,  66,  67,
        68,  69,  70,  71,  72,  75,  76,  77,  80,  81,  82,  83,  84,
        85,  87,  90,  91,  92,  93,  94,  95,  96,  97,  98, 101, 102,
        16,  42,  41,  17,  22,  15, 104,  47,  57,  10, 105, 106,  28,
       107,  65,  52,  33,  56, 108,  27,  38,  36, 109, 110, 112, 113,
       114, 115, 139, 140, 141, 143, 144, 146, 147, 148, 149, 150, 151,
       152, 153, 156, 161, 162, 163, 164, 165, 167, 168, 169, 170, 172,
       173, 174, 175, 177, 178, 179,  29,  19, 180,   6,  99,  43,  31,
         7, 181, 138,  35, 142,   5, 155, 154,  88, 209, 210,  18,   3,
        20, 118, 211, 212, 214, 216, 217, 131, 135, 128, 127, 124, 130,
       159,  12, 145, 241, 242, 136, 157, 171, 123,  25, 134, 119, 243,
       244, 245, 246, 160, 248, 252, 254, 255, 120, 257, 279, 280, 281,
        14, 253, 188, 186, 283, 184, 215, 278, 183, 204, 193, 195, 197,
       249, 206, 284, 121, 201, 285, 286, 311, 190, 312, 116, 20

In [10]:
# Count how many times each user accessed the curriculum
grad_users.user_id.value_counts()

11     17913
64     16297
53     12329
314     7783
1       7404
       ...  
325        1
593        1
533        1
592        1
163        1
Name: user_id, Length: 592, dtype: int64

In [11]:
# Count how many times each page was accessed
grad_users.page.value_counts()

/                                                               15524
javascript-i                                                     4965
spring                                                           4262
search/search_index.json                                         4174
html-css                                                         3678
                                                                ...  
users/1/edit                                                        1
9_Appendix_TSAD_Lesson2                                             1
Dataset_Challenge                                                   1
2.0_Intro_Stats                                                     1
appendix/professional-development/post-interview-review-form        1
Name: page, Length: 1865, dtype: int64

In [12]:
# List unique programs
grad_users.program_id.unique()

array(['PHP', 'Java', 'Front End', 'Data Science'], dtype=object)

In [13]:
# LOOK AT GRAD USERS
grad_program_views = grad_users.groupby(['program_id'])['page'].agg(['count', 'nunique'])
grad_program_views.sort_values(by='count', ascending=False)

count  nunique
program_id                   
Java          157786     1656
PHP            30548      710
Data Science   11495      437
Front End          5        4

In [14]:
# LOOK AT WHOLE DF
df_program_views = df.groupby(['program_id'])['page'].agg(['count', 'nunique'])
df_program_views.sort_values(by='count', ascending=False)

count  nunique
program_id                   
Java          713365     1913
Data Science  103411      682
nan            52893     1112
PHP            30548      710
Front End          5        4

In [15]:
grad_users.groupby('program_id')['page'].value_counts()

program_id    page                                    
Data Science  /                                           1436
              search/search_index.json                     493
              sql/mysql-overview                           275
              classification/overview                      266
              classification/scale_features_or_not.svg     219
                                                          ... 
PHP           students                                       1
              students/468/notes                             1
              students/units/75/sub_units/268                1
              teams/13                                       1
              uploads/58a217a705bde.jpg                      1
Name: page, Length: 2807, dtype: int64

# Data Science

In [16]:
# Create DataFrame for Data Science users that have accessed the curriculum past there graduation date
ds_grad_users = grad_users[grad_users.program_id == 'Data Science']

In [17]:
ds_grad_users.T

date                 2020-01-31                                2020-01-31  \
date        2020-01-31 00:00:00                       2020-01-31 00:00:00   
time                   11:05:04                                  11:05:13   
page                          /  1-fundamentals/1.1-intro-to-data-science   
user_id                     476                                       476   
cohort_id                  34.0                                      34.0   
ip                136.50.49.145                             136.50.49.145   
name                      Bayes                                     Bayes   
start_date  2019-08-19 00:00:00                       2019-08-19 00:00:00   
end_date    2020-01-30 00:00:00                       2020-01-30 00:00:00   
program_id         Data Science                              Data Science   

date                                      2020-01-31  \
date                             2020-01-31 00:00:00   
time                                        11:05:13   
page        1-fundamentals/modern-data-scientist.jpg   
user_id                                          476   
cohort_id                                       34.0   
ip                                     136.50.49.145   
name                                           Bayes   
start_date                       2019-08-19 00:00:00   
end_date                         2020-01-30 00:00:00   
program_id                              Data Science   

date                                  2020-01-31           2020-01-31  \
date                         2020-01-31 00:00:00  2020-01-31 00:00:00   
time                                    11:05:13             14:44:59   
page        1-fundamentals/AI-ML-DL-timeline.jpg                    /   
user_id                                      476                  476   
cohort_id                                   34.0                 34.0   
ip                                 136.50.49.145        136.50.49.145   
name                                       Bayes                Bayes   
start_date                   2019-08-19 00:00:00  2019-08-19 00:00:00   
end_date                     2020-01-30 00:00:00  2020-01-30 00:00:00   
program_id                          Data Science         Data Science   

date                      2020-01-31                          2020-01-31  \
date             2020-01-31 00:00:00                 2020-01-31 00:00:00   
time                        14:47:15                            14:47:20   
page        search/search_index.json  4-python/7.4.4-advanced-dataframes   
user_id                          476                                 476   
cohort_id                       34.0                                34.0   
ip                     136.50.49.145                       136.50.49.145   
name                           Bayes                               Bayes   
start_date       2019-08-19 00:00:00                 2019-08-19 00:00:00   
end_date         2020-01-30 00:00:00                 2020-01-30 00:00:00   
program_id              Data Science                        Data Science   

date                    2020-01-31                  2020-01-31  \
date           2020-01-31 00:00:00         2020-01-31 00:00:00   
time                      14:47:26                    14:47:27   
page        3-sql/1-mysql-overview  3-sql/2-mysql-introduction   
user_id                        476                         476   
cohort_id                     34.0                        34.0   
ip                   136.50.49.145               136.50.49.145   
name                         Bayes                       Bayes   
start_date     2019-08-19 00:00:00         2019-08-19 00:00:00   
end_date       2020-01-30 00:00:00         2020-01-30 00:00:00   
program_id            Data Science                Data Science   

date                 2020-01-31  ...                  2021-04-21  \
date        2020-01-31 00:00:00  ...         2021-04-21 00:00:00   
time                   14:47:28

In [18]:
# Count how many times each page was accessed
ds_grad_users.page.value_counts().head(10)

/                                                 1436
search/search_index.json                           493
sql/mysql-overview                                 275
classification/overview                            266
classification/scale_features_or_not.svg           219
anomaly-detection/AnomalyDetectionCartoon.jpeg     193
anomaly-detection/overview                         191
fundamentals/AI-ML-DL-timeline.jpg                 189
fundamentals/modern-data-scientist.jpg             187
fundamentals/intro-to-data-science                 184
Name: page, dtype: int64

In [19]:
# Look at curriculum by click count and student nunique
ds_program_views = ds_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
ds_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                          
/                                                1436       57
search/search_index.json                          493       42
sql/mysql-overview                                275       38
classification/overview                           266       39
classification/scale_features_or_not.svg          219       36
anomaly-detection/AnomalyDetectionCartoon.jpeg    193       31
anomaly-detection/overview                        191       31
fundamentals/AI-ML-DL-timeline.jpg                189       36
fundamentals/modern-data-scientist.jpg            187       35
fundamentals/intro-to-data-science                184       36

## Check the min/max Graduation Dates.

In [20]:
# Create Java for DS users
DS_users = df[df.program_id == 'Data Science']

In [21]:
# show max end date for DS users
DS_users.end_date.min()

Timestamp('2020-01-30 00:00:00')

In [22]:
# show max end date for DS users
DS_users.end_date.max()

Timestamp('2021-09-03 00:00:00')

In [23]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

# PHP

In [24]:
# Create DataFrame for PHP users that have accessed the curriculum past there graduation date
php_grad_users = grad_users[grad_users.program_id == 'PHP']

In [25]:
php_grad_users.T

date                 2018-01-26           2018-01-26  \
date        2018-01-26 00:00:00  2018-01-26 00:00:00   
time                   09:55:03             09:56:02   
page                          /              java-ii   
user_id                       1                    1   
cohort_id                   8.0                  8.0   
ip                 97.105.19.61         97.105.19.61   
name                    Hampton              Hampton   
start_date  2015-09-22 00:00:00  2015-09-22 00:00:00   
end_date    2016-02-06 00:00:00  2016-02-06 00:00:00   
program_id                  PHP                  PHP   

date                                 2018-01-26  \
date                        2018-01-26 00:00:00   
time                                   09:56:05   
page        java-ii/object-oriented-programming   
user_id                                       1   
cohort_id                                   8.0   
ip                                 97.105.19.61   
name                                    Hampton   
start_date                  2015-09-22 00:00:00   
end_date                    2016-02-06 00:00:00   
program_id                                  PHP   

date                                2018-01-26           2018-01-26  \
date                       2018-01-26 00:00:00  2018-01-26 00:00:00   
time                                  09:56:06             10:14:47   
page        slides/object_oriented_programming                    /   
user_id                                      1                   11   
cohort_id                                  8.0                  1.0   
ip                                97.105.19.61         97.105.19.61   
name                                   Hampton               Arches   
start_date                 2015-09-22 00:00:00  2014-02-04 00:00:00   
end_date                   2016-02-06 00:00:00  2014-04-22 00:00:00   
program_id                                 PHP                  PHP   

date                 2018-01-26              2018-01-26  \
date        2018-01-26 00:00:00     2018-01-26 00:00:00   
time                   10:14:51                10:15:02   
page               javascript-i  javascript-i/functions   
user_id                      11                      11   
cohort_id                   1.0                     1.0   
ip                 97.105.19.61            97.105.19.61   
name                     Arches                  Arches   
start_date  2014-02-04 00:00:00     2014-02-04 00:00:00   
end_date    2014-04-22 00:00:00     2014-04-22 00:00:00   
program_id                  PHP                     PHP   

date                    2018-01-26           2018-01-26  \
date           2018-01-26 00:00:00  2018-01-26 00:00:00   
time                      10:40:15             11:26:13   
page        javascript-i/functions               java-i   
user_id                          1                    1   
cohort_id                      8.0                  8.0   
ip                    97.105.19.61         97.105.19.61   
name                       Hampton              Hampton   
start_date     2015-09-22 00:00:00  2015-09-22 00:00:00   
end_date       2016-02-06 00:00:00  2016-02-06 00:00:00   
program_id                     PHP                  PHP   

date                      2018-01-26  ...           2021-04-09  \
date             2018-01-26 00:00:00  ...  2021-04-09 00:00:00   
time                        11:43:17  ...             00:10:29   
page        mkdocs/search_index.json  ...   content/javascript   
user_id                            1  ...                   80   
cohort_id                        8.0  ...                 14.0   
ip                      97.105.19.61  ...        136.50.29.193   
name                         Hampton  ...               Lassen   
start_date       2015-09-22 00:00:00  ...  2016-07-18 00:00:00   
end_date         2016-02-06 00:00:00  ...  2016-11-10 00:00:00   
program_id                       PHP  ...                  PHP   

date        

In [26]:
# Count how many times each page was accessed
php_grad_users.page.value_counts().head(10)

/                1681
index.html       1011
javascript-i      736
html-css          542
spring            501
java-iii          479
java-ii           454
java-i            444
javascript-ii     429
appendix          409
Name: page, dtype: int64

In [27]:
# Look at curriculum by click count and student nunique
php_program_views = php_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
php_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                         
/               1681       90
index.html      1011       20
javascript-i     736       35
html-css         542       40
spring           501       25
java-iii         479       20
java-ii          454       24
java-i           444       40
javascript-ii    429       33
appendix         409       25

## Why are Curriculum access counts the same for PHP during and after graduation?

In [28]:
# Create DataFrame for PHP users
php_users = df[df.program_id == 'PHP']

In [29]:
# show max end date for Front End users
php_users.end_date.max()

Timestamp('2017-09-22 00:00:00')

In [30]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

In [31]:
# Count how many times each page was accessed
php_users.page.value_counts().head(10)

/                1681
index.html       1011
javascript-i      736
html-css          542
spring            501
java-iii          479
java-ii           454
java-i            444
javascript-ii     429
appendix          409
Name: page, dtype: int64

## ANSWER: Because the last of the PHP users graduated in 2017, before the start date of our data access collection which begins in 2018.

# JAVA

In [32]:
# Create DataFrame for Java users that have accessed the curriculum past there graduation date
java_grad_users = grad_users[grad_users.program_id == 'Java']

In [33]:
java_grad_users.T

date                 2018-01-26           2018-01-26           2018-01-26  \
date        2018-01-26 00:00:00  2018-01-26 00:00:00  2018-01-26 00:00:00   
time                   10:59:02             10:59:06             11:36:27   
page                 index.html               java-i                    /   
user_id                      21                   21                   26   
cohort_id                  16.0                 16.0                 18.0   
ip                71.137.146.71        71.137.146.71         97.105.19.61   
name                    Niagara              Niagara            Pinnacles   
start_date  2016-10-26 00:00:00  2016-10-26 00:00:00  2017-03-27 00:00:00   
end_date    2017-03-09 00:00:00  2017-03-09 00:00:00  2017-07-20 00:00:00   
program_id                 Java                 Java                 Java   

date                 2018-01-26              2018-01-26  \
date        2018-01-26 00:00:00     2018-01-26 00:00:00   
time                   11:36:35                11:36:38   
page               javascript-i  javascript-i/functions   
user_id                      26                      26   
cohort_id                  18.0                    18.0   
ip                 97.105.19.61            97.105.19.61   
name                  Pinnacles               Pinnacles   
start_date  2017-03-27 00:00:00     2017-03-27 00:00:00   
end_date    2017-07-20 00:00:00     2017-07-20 00:00:00   
program_id                 Java                    Java   

date                               2018-01-26           2018-01-29  \
date                      2018-01-26 00:00:00  2018-01-29 00:00:00   
time                                 13:57:51             09:28:24   
page        spring/extra-features/file-upload               jquery   
user_id                                    40                   40   
cohort_id                                16.0                 16.0   
ip                               97.105.19.61         97.105.19.61   
name                                  Niagara              Niagara   
start_date                2016-10-26 00:00:00  2016-10-26 00:00:00   
end_date                  2017-03-09 00:00:00  2017-03-09 00:00:00   
program_id                               Java                 Java   

date                 2018-01-29                          2018-01-29  \
date        2018-01-29 00:00:00                 2018-01-29 00:00:00   
time                   09:28:31                            09:31:01   
page                jquery/ajax  jquery/ajax/requests-and-responses   
user_id                      40                                  40   
cohort_id                  16.0                                16.0   
ip                 97.105.19.61                        97.105.19.61   
name                    Niagara                             Niagara   
start_date  2016-10-26 00:00:00                 2016-10-26 00:00:00   
end_date    2017-03-09 00:00:00                 2017-03-09 00:00:00   
program_id                 Java                                Java   

date                       2018-01-29  ...                 2021-04-21  \
date              2018-01-29 00:00:00  ...        2021-04-21 00:00:00   
time                         10:20:35  ...                   16:07:17   
page        javascript-i/conditionals  ...  javascript-ii/RESTful-api   
user_id                            26  ...                         64   
cohort_id                        18.0  ...                       28.0   
ip                       97.105.19.61  ...              71.150.217.33   
name                        Pinnacles  ...                      Staff   
start_date        2017-03-27 00:00:00  ...        2014-02-04 00:00:00   
end_date          2017-07-20 00:00:00  ...        2014-02-04 00:00:00   
program_id                       Java  ...                       Java   

date                 2021-04-21           2021-04-21           2021-04-21  \
date        2021-04-21 00:00:00  2021-04-21 00:00:00  2021-04-21 00:

In [34]:
# Count how many times each page was accessed
java_grad_users.page.value_counts().head(10)

/                           12406
javascript-i                 4229
spring                       3760
search/search_index.json     3562
html-css                     3136
java-iii                     3058
java-ii                      2985
java-i                       2679
appendix                     2662
javascript-ii                2549
Name: page, dtype: int64

In [35]:
# Look at curriculum by click count and student nunique
java_program_views = java_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
java_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                    
/                         12406      406
javascript-i               4229      300
spring                     3760      306
search/search_index.json   3562      262
html-css                   3136      296
java-iii                   3058      273
java-ii                    2985      274
java-i                     2679      300
appendix                   2662      279
javascript-ii              2549      264

## Check the min/max Graduation Dates.

In [36]:
# Create Java for PHP users
java_users = df[df.program_id == 'Java']

In [37]:
# show max end date for Java users
java_users.end_date.min()

Timestamp('2014-02-04 00:00:00')

In [38]:
# show max end date for Java users
java_users.end_date.max()

Timestamp('2021-10-01 00:00:00')

In [39]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

# Front End

In [40]:
# Create DataFrame for Front End users that have accessed the curriculum past there graduation date
fe_grad_users = grad_users[grad_users.program_id == 'Front End']

In [41]:
fe_grad_users.T

date                 2018-03-22           2018-03-22  \
date        2018-03-22 00:00:00  2018-03-22 00:00:00   
time                   19:01:49             19:01:54   
page                          /     content/html-css   
user_id                     152                  152   
cohort_id                   9.0                  9.0   
ip                207.68.209.17        207.68.209.17   
name                     Apollo               Apollo   
start_date  2015-03-30 00:00:00  2015-03-30 00:00:00   
end_date    2015-07-29 00:00:00  2015-07-29 00:00:00   
program_id            Front End            Front End   

date                                         2018-03-22           2018-03-22  \
date                                2018-03-22 00:00:00  2018-03-22 00:00:00   
time                                           19:01:54             19:02:47   
page        content/html-css/gitbook/images/favicon.ico     content/html-css   
user_id                                             152                  152   
cohort_id                                           9.0                  9.0   
ip                                        207.68.209.17        207.68.209.17   
name                                             Apollo               Apollo   
start_date                          2015-03-30 00:00:00  2015-03-30 00:00:00   
end_date                            2015-07-29 00:00:00  2015-07-29 00:00:00   
program_id                                    Front End            Front End   

date                                2018-03-22  
date                       2018-03-22 00:00:00  
time                                  19:02:52  
page        content/html-css/introduction.html  
user_id                                    152  
cohort_id                                  9.0  
ip                               207.68.209.17  
name                                    Apollo  
start_date                 2015-03-30 00:00:00  
end_date                   2015-07-29 00:00:00  
program_id                           Front End

In [42]:
# Count how many times each page was accessed
fe_grad_users.page.value_counts().head(10)

content/html-css                               2
/                                              1
content/html-css/gitbook/images/favicon.ico    1
content/html-css/introduction.html             1
Name: page, dtype: int64

In [43]:
# Look at curriculum by click count and student nunique
fe_program_views = fe_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
fe_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                       
content/html-css                                 2        1
/                                                1        1
content/html-css/gitbook/images/favicon.ico      1        1
content/html-css/introduction.html               1        1

<div class="alert alert-info">

    
# Q#6 
# ANSWER:

## 1. Data Science
    
|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|
|sql/mysql-overview|275|38|
|classification/overview|266|39|
|anomaly-detection/overview|191|31|
|fundamentals/intro-to-data-science|184|36|

## 2. PHP

|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|
|javascript-i	|736|	35|
|html-css	|542|	40|
|spring	|501|	25|
|java-iii	|479|	20|
    
## 3. Java
    
|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|    
|javascript-i	|4229|	300|
|spring	|3760|	306|
|html-css	|3136|	296|
|java-iii	|3058|	273|
    
## 4. Front End    
    
|Curriculum        |Click Count|Student Count|
|:-----------------|:-:|:-:|    
|content/html-css	|2	|1|
|content/html-css/introduction.html	|1	|1|    

In [48]:
# Look at curriculum by click count and student nunique
ds_program_views = ds_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
ds_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                          
/                                                1436       57
search/search_index.json                          493       42
sql/mysql-overview                                275       38
classification/overview                           266       39
classification/scale_features_or_not.svg          219       36
anomaly-detection/AnomalyDetectionCartoon.jpeg    193       31
anomaly-detection/overview                        191       31
fundamentals/AI-ML-DL-timeline.jpg                189       36
fundamentals/modern-data-scientist.jpg            187       35
fundamentals/intro-to-data-science                184       36

In [49]:
# Look at curriculum by click count and student nunique
php_program_views = php_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
php_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                         
/               1681       90
index.html      1011       20
javascript-i     736       35
html-css         542       40
spring           501       25
java-iii         479       20
java-ii          454       24
java-i           444       40
javascript-ii    429       33
appendix         409       25

In [50]:
# Look at curriculum by click count and student nunique
java_program_views = java_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
java_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                    
/                         12406      406
javascript-i               4229      300
spring                     3760      306
search/search_index.json   3562      262
html-css                   3136      296
java-iii                   3058      273
java-ii                    2985      274
java-i                     2679      300
appendix                   2662      279
javascript-ii              2549      264

In [51]:
# Look at curriculum by click count and student nunique
fe_program_views = fe_grad_users.groupby(['page'])['user_id'].agg(['count', 'nunique'])
fe_program_views.sort_values(by='count', ascending=False).head(10)

count  nunique
page                                                       
content/html-css                                 2        1
/                                                1        1
content/html-css/gitbook/images/favicon.ico      1        1
content/html-css/introduction.html               1        1

<div class="alert alert-warning">


## 7. Which lessons are least accessed?

## Why are there so few Front End Graduates accessing the Curriculum?

In [44]:
# Create DataFrame for Front End users
fe_users = df[df.program_id == 'Front End']

In [45]:
# show max end date for Front End users
fe_users.end_date.max()

Timestamp('2015-07-29 00:00:00')

In [46]:
# show min access date for df
df.date.min()

Timestamp('2018-01-26 00:00:00')

In [47]:
# How many users and cohorts make up the Front End
fe_users.groupby(['name'])['user_id'].agg(['nunique'])

nunique
name           
Apollo        1

<div class="alert alert-info">


## ANSWER: Because there was only one Front End user who graduated in 2015, three years before the start date of our data access collection which begins in 2018.

## Most likely, the majority of the Null entries for Cohort could probably be attributed to Front End as well as a few towards Staff

## 8. Anything else I should be aware of?